protocol/x/ratelimit/keeper/packet.go

```diff
@@ -0,0 +1,181 @@
+package keeper
+
+// TODO(CORE-854): Improve attribution message.
+// This file was written with reference to Stride's IBC Rate Limit implementation:
+// https://github.com/Stride-Labs/stride/blob/121f2ac5d2e5f8e406f89999410a49ea4277a552/x/ratelimit
+
+import (
+	"fmt"
+	"math/big"
+
+	"cosmossdk.io/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	capabilitytypes "github.com/cosmos/ibc-go/modules/capability/types"
+	clienttypes "github.com/cosmos/ibc-go/v8/modules/core/02-client/types" //nolint:staticcheck
+	channeltypes "github.com/cosmos/ibc-go/v8/modules/core/04-channel/types"
+	ibcexported "github.com/cosmos/ibc-go/v8/modules/core/exported"
+	"github.com/dydxprotocol/v4-chain/protocol/x/ratelimit/types"
+	"github.com/dydxprotocol/v4-chain/protocol/x/ratelimit/util"
+)
+
+// Remove a pending packet sequence number from the store
+// Used after the ack or timeout for a packet has been received
+func (k Keeper) RemovePendingSendPacket(ctx sdk.Context, channelId string, sequence uint64) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), []byte(types.PendingSendPacketPrefix))
+	key := types.GetPendingSendPacketKey(channelId, sequence)
+	store.Delete(key)
+}
+
+// Sets the sequence number of a packet that was just sent
+func (k Keeper) SetPendingSendPacket(ctx sdk.Context, channelId string, sequence uint64) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), []byte(types.PendingSendPacketPrefix))
+	key := types.GetPendingSendPacketKey(channelId, sequence)
+	store.Set(key, []byte{1}) // Use a single bit to indicate packet is pending.
+}
+
+// Checks whether the packet sequence number is in the store - indicating that it is a pending
+// packet.
+func (k Keeper) HasPendingSendPacket(ctx sdk.Context, channelId string, sequence uint64) bool {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), []byte(types.PendingSendPacketPrefix))
+	key := types.GetPendingSendPacketKey(channelId, sequence)
+
+	return store.Has(key)
+}
+
+// Middleware implementation for OnAckPacket
+// It is called on the sender chain when a relayer relays back the acknowledgement from the receiver chain.
+// On the dYdX chain, this includes the “response” of the receiver chain for outbound transfer from dYdX.
+func (k Keeper) AcknowledgeIBCTransferPacket(
```

Comment:
`Ack` is the final step of an IBC transfer - see full life cycle [here](https://tutorials.cosmos.network/academy/3-ibc/7-token-transfer.html#transfer-packet-flow).

If the acknowledgement indicates a failure on the receiver chain then we revert the capacity decrease. Otherwise, we simply complete the flow by removing "pending" packet. 

Similar to `OnRecvPacket`, `Acknowledge` is called when relayer submits an acknowledgement as part of a transaction. All state changes are committed only if the underlying `Acknowledge` logic of the transfer application succeeds as well